proposals/0008-package-manager-testing.md

```diff
@@ -0,0 +1,194 @@
+# Swift Testing
+
+* Authors:
+  [Max Howell](https://github.com/mxcl),
+  [Daniel Dunbar](https://github.com/ddunbar),
+  [Mattt Thompson](https://github.com/mattt)
+* Status: **Review**
+* Review Manager: Rick Ballard
+
+## Introduction
+
+Testing is an essential part of modern software development.
+Tight integration of testing into the Swift Package Manager
+will help ensure a stable and reliable packaging ecosystem.
+
+## Proposed Solution
+
+We propose to extend our conventional package directory layout
+to accomodate test modules.
+Any subdirectory of the package root directory named "Tests"
+or any subdirectory of an existing module directory named "Tests"
+will comprise a test module.
+For example:
+
+    Package
+    ├── Sources
+    │   └── Foo
+    │       └──Foo.swift
+    └── Tests
+        └── Foo
+            └── Test.swift
+
+Or:
+
+    Package
+    └── Sources
+        ├── Foo.swift
+        └── Tests
+            └── Test.swift
+
+Or, for simpler projects:
+
+    Package
+    ├── Sources
+    │   └── Foo.swift
+    └── Tests
+        └── TestFoo.swift
+        
+The filename: `TestFoo.swift` is arbitrary.
+
+In the examples above
+a test case is created for the module `Foo`
+based on the sources in the relevant subdirectories.
+
+Additionally we will support directories called `FooTests`.
+This layout style is prevalent in existing open source projects
+and supporting it will minimize vexation for their authors.
+However in the interest of consistency and the corresponding 
+reduction of cognitive-load when examining new Swift packages
+we will not recommend this layout. For example:
+
+    Package
+    └── Sources
+    │   └── Foo.swift
+    └── FooTests
+        └── Test.swift
+
+Additionally, we propose that building a module
+also builds that module's corresponding tests.
```

Comment:
Excellent idea! But would it be possible to explicitly disable this behavior by passing command-line options to swiftpm?

I ask because I can envision several situations in which this is not desirable behavior, such as when the tests are known to be broken. In this situation, a developer could pull the latest revision down because its tests are failing on a continuous integration suite, and needs to then build and run that revision in order to attach a debugger like LLDB.